labrad/oauth.py

```diff
@@ -0,0 +1,284 @@
+
+    Args:
+        client_id (str): The client_id for the app we want to authorize.
+        client_secret (str): The client_secret for the app we want to authorize.
+        headless (bool): If False, the default, we open a web browser to perform
+            the OAuth login and spin up a local http server to receive the
+            authorization code. If True, or if we fail to open a webbrowser,
+            we print a url for the user to visit to get the authorization code,
+            which they must then paste into the terminal. The headless is useful
+            in some scenarios such as when tunneling through ssh.
+        retries (int): The number of times to retry picking a random port for
+            the local http server.
+        timeout (int): Amount of time to wait for OAuth login flow to complete,
+            in seconds.
+
+    Returns:
+        (str) id_token to be passed to the manager to identify the user.
+    """
+    cached = _load_token(client_id)
+    if cached is None:
+        print('No cached token. Logging in...')
+    else:
+        if cached['expires_at'] > time.time() + 600:
+            print('Using cached id_token...')
+            return OAuthToken(cached)
+        print('Trying cached refresh_token...')
+        try:
+            token = _refresh_token(client_id, client_secret,
+                                   cached['refresh_token'])
+            _cache_token(client_id, token)
+            return OAuthToken(token)
+        except Exception:
+            print('Failed to refresh token. Logging in...')
+
+    def do_headless_login():
+        redirect_uri = NO_REDIRECT_URI
+        login_uri = _create_login_uri(client_id, redirect_uri)
+        print('To obtain an OAuth authorization code, please navigate to the '
+              'following URL in a browser:\n\n{}\n'.format(login_uri))
+        code = raw_input('When you have completed the login flow, enter the '
+                         'code here: ')
+        return redirect_uri, code
+
+    if headless:
+        redirect_uri, code = do_headless_login()
+    else:
+        code_future = futures.Future()
+
+        class OAuthHandler(BaseHTTPServer.BaseHTTPRequestHandler):
+            def do_GET(self):
+                parsed_path = urlparse.urlparse(self.path)
+                params = urlparse.parse_qs(parsed_path.query)
+                code = params['code'][0]
+                code_future.set_result(code)
+                self.send_response(200)
+                self.send_header('Content-Type', 'text/html')
+                self.end_headers()
+                self.wfile.write(_OAUTH_PAGE)
+
+            def log_request(self, *args, **kw):
+                pass
+
+        for i in xrange(retries):
+            local_port = random.randrange(10000, 50000)
+            try:
+                httpd = BaseHTTPServer.HTTPServer(('127.0.0.1', local_port),
+                                                  OAuthHandler)
```

Comment:
use "localhost" instead of 127.0.0.1 for IPv6 compatibility.

If you just use local_port=0, it will assign an unused high port automatically, which you can retrieve with: `httpd.server_address` and avoid the looping (which would be undesirable if the error were something other than port in use, especially if the error is associated with a lengthy timeout)